### DISCUSSION TITLE
Needing advices about programming on LLVM or CUDA variant in C++.

### DISCUSSION THREAD

**Discussion by g1r4ff3** [mitsuba-renderer/mitsuba3#964] [Q&A]
Hello, I need a help with programming on LLVM or CUDA variant for extending existing my code in scalar varaint to those variants. I've written the code in C++ version of Mitsuba 3.
As my attempt to resolve my problem, I've read ["First steps" guide of Dr.Jit](https://drjit.readthedocs.io/en/latest/firststeps-py.html) by [the previous answer](https://github.com/mitsuba-renderer/mitsuba3/discussions/898) and followed the simple solution by setting record features off before my code, but the problem hasn't been resolved. It seems to say that i need to write the code without conventional if-statements in those variants, to delay the evaluation of computation until the images are generated as the document says. But, how could i do that in complex programming situations like coding a quartic equation solver, which varies the way it computes the roots according to various conditions?
I've tried to use the Mask variable to control the execution, but it can't vary the creation of variables relevant to the solver.
Sorry for my poor question, but are there other ways or advices for me to resolve the problem i have?
Thank you in advance.

**Comment by njroussel** [comment]
Hi @g1r4ff3 

Writing this sort of vectorized/SIMD code is a bit unusual. Masking is indeed the key to workaround the traditional if/else constructs. However, when replacing a conditional branch with some branching, both code paths become valid and you therefore **need** to create all variables for both paths.
For your specific example, we actually have a [quadratic equation solver](https://github.com/mitsuba-renderer/mitsuba3/blob/master/include/mitsuba/core/math.h#L357-L401). I hope this helps you to better understand this programming model.

**Reply by g1r4ff3** [reply]
Thank you @njroussel for your kind reply.
Okay, i understand what you said, but i have some subsequent questions which i can't understand.
1. Why can't i use dr::any() or dr::all to iterate through in vectorized variants?
    If i use those in loops, the placeholder error occurs.
2. How can i access and modify an element of dr::Array<LLVMArray\<float\> or CUDAArray\<float\>, N> using integer array like a dr::int32_array<LLVMArray\<float\> or CUDAArray\<float\>> for indexing the dr::Array, with optional masking?
    it seems that the scatter/gather operations would not help me if i understood them well.

**Reply by njroussel** [reply]
1) These operations are horizontal reductions, they will force a kernel evaluation and that is  not allowed in recorded contexts.

2) Gather/scatter operations are indeed what you should be using for indexing operations. However, reading (`gather`) from a variable that was written to (`scatter`) will most likely require a kernel evaluation.
